server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1299,7 +1314,17 @@ protected final DeletionStrategy planDeletionAsPrimary(Delete delete) throws IOE
             currentlyDeleted = versionValue.isDelete();
         }
         final DeletionStrategy plan;
-        if (delete.versionType().isVersionConflictForWrites(currentVersion, delete.version(), currentlyDeleted)) {
+        if (delete.getCasSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO && versionValue == null) {
```

Comment:
I looked at this and the tricky part is that we create a different strategy class as a result. I think it will just make it harder to read if we have a common strategy base class or some callback based code